text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,172 @@
+
+### Filesystem
+
+The filesystem scheme is the implementation of [RFC 8089](https://tools.ietf.org/html/rfc8089).
+
+### URL
+
+The URL scheme is the implementation of [RFC 7230](https://tools.ietf.org/html/rfc7230#section-2.7.1).
+
+### Docker
+
+The docker scheme denotes the use of the [Docker HTTP API v2 protocol](https://docs.docker.com/registry/spec/api/).
+
+- `<host>` is optional and defaults to `index.docker.io`.
+- Similar to the `file` scheme, there is a minimal declaration for omitting host by using a simple slash (`/`). 
+eg. `docker:/ubuntu:latest`
+
+### CNB Builder
+
+- `cnb:builder://` - A reference to ALL the buildpacks in the builder.
+- `cnb:builder:///bp.id` - A reference to a buildpack with id `bp.id` in the builder. If there is more than one version
+available, a error should be thrown.
+- `cnb:builder:///bp.id:bp.version` - A reference to a buildpack with id `bp.id` at the specific version `bp.version`.
+- Similar to the `file` scheme, there is a minimal declaration for omitting host by using a simple slash (`/`). 
+eg. `cnb:builder:/bp.id:bp.version`
+
+# How it Works
+[how-it-works]: #how-it-works
+
+### pack
+
+Certain `pack` CLI arguments would take URIs supported instead of the current inconsistent format.
+
+Some examples:
+
+```shell script
+pack build my-app \
+  --buildpack cnb:builder://bp.id@bp.version \                # buildpack from builder
+  --buildpack file:///home/user/path/to/buildpack/ \          # absolute via file
+  --buildpack /home/user/path/to/buildpack/ \                 # absolute via schemeless
+  --buildpack ../path/to/buildpack/ \                         # relative file
+  --buildpack docker:/cnbs/some-package \                     # Docker Hub image
+  --buildpack docker://gcr.io/cnbs/sample-package-2:bionic \  # GCR image (with tag)
+  --buildpack cnb:builder://                                  # All buildpacks in builder
+```   
+
+### Config Files
+
+Config files would replace `image` with `uri`, using the `docker://` scheme, where applicable.
+
+`package.toml`:
+```toml
+[buildpack]
+uri = "."
+
+[[dependencies]]
+uri = "docker:/my/image:latest"
+
+# ...
+``` 
+
+`builder.toml`:
+```toml
+# ...
+
+[[buildpacks]]
+uri = "./my/buildpacks/"
+
+[[buildpacks]]
+uri = "docker:/my/image:latest"
+
+# ...
+```
+
+# Additional considerations
+
+#### `pack` User Expirience
+
+`pack` could provide additional logic in order for users to have a nicer user experience when providing references
+via CLI. For example, `pack` could have an order of precedence to make a best guess at which type of reference
+they are referring to or provide additional feedback if it's ambiguous.
+
+#### Registry vs Daemon
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The user would need to type an additional scheme prefix for any URI which is not a relative reference.
+
+# Alternatives
```

Comment:
As an alternative, what if we defined `cnb` URN patterns. Unlike URLs, which identify resources by specifying their locations, URNs are location-independent identifiers assigned within defined namespaces.

 For example:

```
# reference a buildpack in the builder
urn:cnb:builder:[<id>[:<version>]]

# reference a buildpack on the default cnb-registry
urn:cnb:registry:[<id>[:<version>]]
```

In this way, the Builder image or the Registry location can be defined independently. I think this will be pretty common (`pack` already has a default builder, and will have a default registry).

If you need to specify buildpacks in more than one registry, then use the following URI:

```
# always points to a buildpack on a specific cnb-registry
cnb://[<host>]/[<id>[:<version>]]
```

Because there's no concept of pointing to buildpacks in more than one builder, we don't need to define a URI for buildpacks-in-a-builder (i.e. there's no need to define the location, so the URN should do).

Then we can honor all the other `file://`, `docker://`, etc schemes defined in the RFC.

Reply:
Oooh I like this!

Reply:
Thank you! I've included this suggestion into the latest changes.